src/stats/stats_mempool.cpp

```diff
@@ -0,0 +1,130 @@
+std::atomic<CAmount> CStatsMempool::cacheMempoolMinRelayFee;
+
+CStatsMempool::CStatsMempool(unsigned int collectIntervalIn) : collectInterval(collectIntervalIn)
+{
+    startTime = 0;
+    intervalCounter = 0;
+
+    // setup the samples per precision vector
+    for (unsigned int interval : precisionIntervals) {
+        (void)(interval);
+        vSamplesPerPrecision.emplace_back();
+
+        // use the fallback max in case max memory will not be set
+        vMaxSamplesPerPrecision.push_back(fallbackMaxSamplesPerPrecision);
+
+        // add starttime 0 to each level
+        vTimeLastSample.push_back(0);
+    }
+}
+
+std::vector<unsigned int> CStatsMempool::getPrecisionGroupsAndIntervals() {
+    return {std::begin(precisionIntervals), std::end(precisionIntervals)};
+}
+
+bool CStatsMempool::addMempoolSamples(const size_t maxStatsMemory)
+{
+    bool statsChanged = false;
+    uint64_t now = GetTime();
+    {
+        LOCK(cs_mempool_stats);
+
+        // set the mempool stats start time if this is the first sample
+        if (startTime == 0)
+            startTime = now;
+
+        unsigned int biggestInterval = 0;
+        for (unsigned int i = 0; i < sizeof(precisionIntervals) / sizeof(precisionIntervals[0]); i++) {
+            // check if it's time to collect a samples for the given precision level
+            uint16_t timeDelta = 0;
+            if (intervalCounter % (precisionIntervals[i] / (collectInterval / 1000)) == 0) {
+                if (vTimeLastSample[i] == 0) {
+                    // first sample, calc delta to starttime
+                    timeDelta = now - startTime;
+                } else {
+                    timeDelta = now - vTimeLastSample[i];
+                }
+                vSamplesPerPrecision[i].push_back({timeDelta, CStatsMempool::cacheMempoolSize, CStatsMempool::cacheMempoolDynamicMemoryUsage, CStatsMempool::cacheMempoolMinRelayFee});
+                statsChanged = true;
+
+                // check if we need to remove items at the beginning
+                if (vSamplesPerPrecision[i].size() > vMaxSamplesPerPrecision[i]) {
+                    // increase starttime by the removed deltas
+                    for (unsigned int j = (vSamplesPerPrecision[i].size() - vMaxSamplesPerPrecision[i]); j > 0; j--) {
+                        startTime += vSamplesPerPrecision[i][j].timeDelta;
+                    }
+                    // remove element(s) at vector front
+                    vSamplesPerPrecision[i].erase(vSamplesPerPrecision[i].begin(), vSamplesPerPrecision[i].begin() + (vSamplesPerPrecision[i].size() - vMaxSamplesPerPrecision[i]));
+
+                    // release memory
+                    vSamplesPerPrecision[i].shrink_to_fit();
+                }
+
+                vTimeLastSample[i] = now;
+            }
+            biggestInterval = precisionIntervals[i];
+        }
+
+        intervalCounter++;
+
+        if (intervalCounter > biggestInterval) {
```

Comment:
In commit "Add mempool statistics collector"

I don't understand what this is doing. It seems unnecessary, and strange because it is comparing the counter against numbers of seconds without taking collectInterval into account. If you will keep this, I think it'd be good to add a comment with an example of interval and counter values that explains how this is supposed to work.